### DISCUSSION TITLE
Is it possible to indent headings?

### DISCUSSION THREAD

**Discussion by kurt-rhee** [typst/typst#2812] [Q&A] [answered]
Hello, 

Apologies if this has been answered elsewhere, I was wondering if it is possible to indent the headings by some amount by the level of the heading in the heading hierarchy.  These headings are not part of an outline.

![image](https://github.com/typst/typst/assets/33131958/d476bf9d-4f9c-4e69-84e0-9bd33da902bc)

**Comment by frozolotl** [comment, marked as answer]
Indenting just the headings is pretty easy:
```typ
#show heading: it => pad(left: 1em * it.level, it)

= H1
#lorem(5)

== H2
#lorem(5)
```

I'm not sure this is what you want though, as this does not indent anything below the headings.
The following solution requires you to manually use these `indent` function calls:
```typ
#let indent = pad.with(left: 1em)

= H1
#lorem(5)

#indent[
  == H2
  #lorem(5)

  #indent[
    === H3
    #lorem(5)
  ]
]
```

**Reply by astrojhgu** [reply]
With the `pad` function, the whole heading paragraph are shifted. What if I only want the first line to be indented?

**Reply by frozolotl** [reply]
You can use `#set par(first-line-indent: 1em)` for that, iirc.
Though note that this would apply only to consecutive paragraphs at the moment.

**Reply by astrojhgu** [reply]
I did add `#set par (first-line-indent: 2em)` in the `#show heading` section, but nothing happened. 

![image](https://github.com/typst/typst/assets/2651231/144ebcf7-54c6-4445-8ce2-24f6bfd66e5f)


```typst
//I have a complex heading numbering.

#let clean_numbering(..schemes) = {
  (..nums) => {
    let (section, ..subsections) = nums.pos()
    let (section_scheme, ..subschemes) = schemes.pos()

    if subsections.len() == 0 {
      numbering(section_scheme, section)
    } else if subschemes.len() == 0 {
      numbering(section_scheme, ..nums.pos())
    }
    else {
      clean_numbering(..subschemes)(..subsections)
    }
  }
}


#set heading(numbering:clean_numbering("I","1.","a"))


#show heading.where(
  level: 1
): it => block(width: 100%)[
  #set text(16pt, weight: "regular",fill:blue)
  #set par(first-line-indent: 2em) //this line does not change anything
  #it
]

#show heading.where(
  level: 2
): it => block(width: 100%)[
  #set text(16pt, weight: "regular",fill:blue)
  #set par(first-line-indent: 1em) //this line does not change anything
  #it
]


#set text(size:15pt)
#set par(first-line-indent: 2em)

= H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 $<-$ I need this line to be indented
#h(2em) Some paragraph

Some paragraph
== H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2  $<-$ and this line
#h(2em) Some paragraph

Some paragraph
= H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 H1 $<-$ and this line
#h(2em) Some paragraph

Some paragraph
== H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2 H2  again $<-$ and this line
#h(2em) Some paragraph

Some paragraph
```

**Reply by frozolotl** [reply]
I think this is getting a bit too large ans out of scope; can you open a new discussion?

**Reply by astrojhgu** [reply]
Thanks. 

I have opened a new [discussion: https://github.com/typst/typst/discussions/3353](https://github.com/typst/typst/discussions/3353).
